### ISSUE TITLE
Image Update Automation not honouring exclusion list

### ISSUE THREAD

**Issue by 2fst4u** [fluxcd/flux2#4210]
I have the following ImageRepository and ImagePolicy applied to my cluster:

```
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageRepository
metadata:
  name: prowlarr
  namespace: flux-system
spec:
  image: lscr.io/linuxserver/prowlarr
  interval: 1h
  exclusionList:
    - "^.*\\.sig$"
    - "1.8.1259"
    - "1.7.1209"
    - "1.6.1196" 
    - "1.5.1164"
    - "1.4.1155"
    - "1.2.1"
---
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImagePolicy
metadata:
  name: prowlarr
  namespace: flux-system
spec:
  imageRepositoryRef:
    name: prowlarr
  filterTags:
    pattern: '^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$'
  policy:
    semver:
      range: '^1.x'
```

My exclusion list contains tags which I know to be false semver for old versions with an implicit "0." at the start, so I have attempted to explicitly ignore them from being used. 

When I describe both my ImageRepository and ImagePolicy it reconciles to tag `1.8.1259` which is the first item in my exclusion list that I specifically don't want:

```
Name:         prowlarr
Namespace:    flux-system
Labels:       kustomize.toolkit.fluxcd.io/name=flux-system
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  image.toolkit.fluxcd.io/v1beta1
Kind:         ImageRepository
Metadata:
  Creation Timestamp:  2023-08-31T02:01:38Z
finalizers.fluxcd.io
  Generation:        1
  Resource Version:  329066117
  UID:               39787277-38c3-4480-b543-f1d455d19630
Spec:
  Exclusion List:
    ^.*\.sig$
    1.8.1259
    1.7.1209
    1.6.1196
    1.5.1164
    1.4.1155
    1.2.1
  Image:     lscr.io/linuxserver/prowlarr
  Interval:  1h
Status:
  Canonical Image Name:  lscr.io/linuxserver/prowlarr
  Conditions:
    Last Transition Time:  2023-08-31T02:01:45Z
    Message:               successful scan, found 65763 tags
    Reason:                ReconciliationSucceeded
    Status:                True
    Type:                  Ready
  Last Scan Result:
    Scan Time:          2023-08-31T02:01:45Z
    Tag Count:          65763
  Observed Generation:  1
Events:                 <none>
```

```
Name:         prowlarr
Namespace:    flux-system
Labels:       kustomize.toolkit.fluxcd.io/name=flux-system
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  image.toolkit.fluxcd.io/v1beta1
Kind:         ImagePolicy
Metadata:
  Creation Timestamp:  2023-08-31T02:01:38Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:        1
  Resource Version:  329066060
  UID:               08a1eed2-5635-4b62-ae56-a43e2ef317e0
Spec:
  Filter Tags:
    Pattern:  ^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
  Image Repository Ref:
    Name:  prowlarr
  Policy:
    Semver:
      Range:  ^1.x
Status:
  Conditions:
    Last Transition Time:  2023-08-31T02:01:38Z
    Message:               Latest image tag for 'lscr.io/linuxserver/prowlarr' resolved to: 1.8.1259
    Reason:                ReconciliationSucceeded
    Status:                True
    Type:                  Ready
  Latest Image:            lscr.io/linuxserver/prowlarr:1.8.1259
  Observed Generation:     1
Events:
  Type     Reason  Age                 From                        Message
  ----     ------  ----                ----                        -------
  Warning  error   101s                image-reflector-controller  ImageRepository.image.toolkit.fluxcd.io "prowlarr" not found
  Normal   info    94s (x3 over 101s)  image-reflector-controller  Latest image tag for 'lscr.io/linuxserver/prowlarr' resolved to: 1.8.1259
```

**Comment by makkes** [comment]
Please paste the output of the following commands here:

1. `flux logs --kind ImageRepository --name prowlarr` (the last couple of lines)
2. `kubectl get imagerepository -n flux-system prowlarr -o yaml`

**Comment by makkes** [comment]
I see you're using API version v1beta1. Is there a specific reason for not using v1beta2 which has been introduced in February this year? Reason I'm asking is that v1beta2 has an extra field ``.status.lastScanResult.latestTags` which v1beta1 does not and that may help us understand what's happening.

> Can I run an equivalent command?

```
kubectl logs -n flux-system -l app=image-reflector-controller | jq 'select(.controllerKind=="ImageRepository")|select(.name=="prowlarr")'
```

**Comment by 2fst4u** [comment]
Because of this part

> it's a long story and I can't do an install and update because I have encrypted secrets in my repo, and last time I tried I broke everything. I have a discussion in this repo about it

Am I able to manually update that manifest to enable beta2?

**Comment by 2fst4u** [comment]
Output of `kubectl logs -n flux-system -l app=image-reflector-controller | jq 'select(.controllerKind=="ImageRepository")|select(.name=="prowlarr")'`

```
{
  "level": "info",
  "ts": "2023-08-31T09:02:39.911Z",
  "msg": "reconciliation finished in 8.537980221s, next run in 1h0m0s",
  "controller": "imagerepository",
  "controllerGroup": "image.toolkit.fluxcd.io",
  "controllerKind": "ImageRepository",
  "ImageRepository": {
    "name": "prowlarr",
    "namespace": "flux-system"
  },
  "namespace": "flux-system",
  "name": "prowlarr",
  "reconcileID": "a7641f0f-c3bf-4827-9a5d-c5a4ce905894"
}
```

**Comment by 2fst4u** [comment]
Ok cool I'll give that a go and report back, cheers.

**Comment by 2fst4u** [comment]
> > Because of this part
> > 
> > > it's a long story and I can't do an install and update because I have encrypted secrets in my repo, and last time I tried I broke everything. I have a discussion in this repo about it
> > 
> > Am I able to manually update that manifest to enable beta2?
> 
> Basically yes. The changelog has all the [upgrade instructions](https://github.com/fluxcd/image-reflector-controller/blob/main/CHANGELOG.md#imagetoolkitfluxcdiov1beta2).

Just catching up to do this now, it seems I won't be able to as my controllers are still on v0.23.0 and not 0.25.0 as needed. I can't upgrade my controllers due to the issue presented above - there isn't a documented supported way for me to do it that won't break my cluster according to the discussion I had.

**Comment by makkes** [comment]
> I can't upgrade my controllers due to the issue presented above

What exactly prevents you from upgrading the controller?

**Comment by 2fst4u** [comment]
https://github.com/fluxcd/flux2/discussions/3443

This discussion outlines my problem.

**Comment by 2fst4u** [comment]
In the absence of the field in the updated API, is there anything else I can look at?

**Comment by 2fst4u** [comment]
Quick thought: is it because the exclusion list was introduced in v1beta2? If so, that's why it's not even working, right?

**Comment by gecube** [comment]
I triaged this one. The good news: it's a real bug, but it was fixed long ago — and the answer to the reporter's last question is "no, but close".

## TL;DR

`exclusionList` **was** part of `v1beta1` (added in image-reflector-controller v0.19.0, May 2022 — a089f86), so the field was valid, persisted and *live* on your cluster. The real problem is a bug in the tag-filtering loop of image-reflector-controller **≤ v0.24.0**: with more than one pattern in `exclusionList`, a tag was excluded only if it matched **every** pattern in the list, not *any* of them. It was fixed (somewhat silently) in the big reconciler refactor fluxcd/image-reflector-controller#311 and first shipped in **image-reflector-controller v0.25.0 = flux v0.40.0 (Feb 2023)** — coincidentally the exact version you were told you'd need, though for a different reason.

## Root cause

In v0.23.0 the filter looked like this ([controllers/imagerepository_controller.go#L375-L386](https://github.com/fluxcd/image-reflector-controller/blob/v0.23.0/controllers/imagerepository_controller.go#L375-L386)):

```go
filteredTags := []string{}
for _, regex := range imageRepo.Spec.ExclusionList {
	r, err := regexp.Compile(regex)
	...
	for _, tag := range tags {
		if !r.MatchString(tag) {
			filteredTags = append(filteredTags, tag)
		}
	}
}
```

For **each** pattern it iterates the **original** tag list and appends every non-matching tag. So with your 7 patterns, `1.8.1259` doesn't match `^.*\.sig$` → appended on the first pass → it's in the database, and the semver policy happily picks it. A tag only ever gets dropped if it matches *all* patterns simultaneously. As a bonus, every surviving tag is stored once per non-matching pattern.

That bonus is visible right in the issue description: `found 65763 tags`. 65763 ≈ **7 × ~9395** (the actual tag count of `lscr.io/linuxserver/prowlarr` in Aug 2023) — one copy per exclusion pattern.

The fix in #311 ([`filterOutTags`](https://github.com/fluxcd/image-reflector-controller/blob/v0.25.0/controllers/imagerepository_controller.go)) compiles all patterns first and keeps a tag only if it matches none of them; the changelog entry even says *"Bug in the exclusion list filter. Multiple patterns in the list didn't filter the tags for all the patterns."*

## Reproduction (today)

Same manifests as in the issue description (exclusion list updated to today's top tags, since prowlarr is at 1.37.x now), against `lscr.io/linuxserver/prowlarr` (15241 tags today):

```yaml
exclusionList:
  - "^.*\\.sig$"
  - "1.37.0"
  - "1.36.3"
  - "1.35.1"
  - "1.34.1"
  - "1.33.3"
  - "1.2.1"
```

| controller | tagCount | policy result (`semver: ^1.x`) |
|---|---|---|
| v0.23.0 (yours, flux v0.37) | **106054** ≈ 7 × 15241 | **`1.37.0`** — the first entry of the exclusion list 🐛 |
| v1.2.3 (current) | **14608** = 15241 − 633 | **`1.32.2`** — highest tag *not* excluded ✅ |

The two runs cross-check exactly: the buggy run stored 7×15241 − 633 = 106054 entries, the fixed run dropped the same 633 matches (15241 − 633 = 14608).

## Two footnotes for anyone landing here

1. The patterns are Go RE2 and matched **unanchored** (substring). `1.2.1` also excludes `1.2.10`, `1.2.1259`, `0.1.2.1-ls34`, … — that's where most of the 633 exclusions above come from. If you want to exclude an exact tag, anchor and escape it: `^1\.8\.1259$`.
2. On current Flux the image APIs are GA (`image.toolkit.fluxcd.io/v1`, only `v1` is served), so the manifests from this issue need an `apiVersion` bump anyway.

Since this has been fixed since flux v0.40.0 and every supported version behaves correctly, this issue can probably be closed.

cc @matheuscscp
